Bug Report: vtgate/buffer — waitForFailoverEnd can start a new buffering cycle after Buffer.Shutdown() has returned

### Overview of the Issue

There is a pre-existing race condition in `go/vt/vtgate/buffer` where a request goroutine that captures a `*shardBuffer` reference via `getOrCreateBuffer` before `Buffer.Shutdown()` completes can subsequently transition that shard into `stateBuffering` _after_ `Buffer.Shutdown()` has already returned. This violates the documented contract on `Buffer.Shutdown`:

> Shutdown blocks until all pending ShardBuffer objects are shut down. In particular, it guarantees that all launched Go routines are stopped after it returns.

The racing call spawns a timeout-thread goroutine (via `startBufferingLocked → timeoutThread.start()`) that is tracked on `timeoutThread.wg`, not on `shardBuffer.wg`, so `Buffer.Shutdown`'s `waitForShutdown` never waits on it.

**Production impact:** not a crash or corruption, but a latency regression. The late caller will block in `sb.wait()` until `MaxFailoverDuration` (default 20s) expires, then the timeout thread triggers `stopBufferingDueToMaxDuration → drain`, which unblocks the request and it retries. Net effect: a query landing in this window sees a very noticeable latency spike before it finally succeeds.

This is **not** a regression from #19827 — it reproduces identically on the PR's base commit. #19827's new `stopped.Load` + `LoadAndDelete` cleanup narrows the window for the cold-start variant (newly-created `sb`) but does not close the window for the case where `sb` already existed in the map. See https://github.com/vitessio/vitess/pull/19827#issuecomment-4288024828 for discussion.

### Reproduction Steps

Drop the following test file into `go/vt/vtgate/buffer/` and run:

```
go test -count=1 -race -run '^TestShutdownRace_LateWaitForFailoverEnd$' -v ./go/vt/vtgate/buffer/...
```

```go
/*
Copyright 2026 The Vitess Authors.
Licensed under the Apache License, Version 2.0 (the "License");
*/

package buffer

import (
	"context"
	"testing"
	"time"

	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

func TestShutdownRace_LateWaitForFailoverEnd(t *testing.T) {
	cfg := NewDefaultConfig()
	cfg.Enabled = true
	cfg.MaxFailoverDuration = 500 * time.Millisecond
	b := New(cfg)

	sb := b.getOrCreateBuffer(keyspace, shard)
	require.NotNil(t, sb)
	require.Equal(t, stateIdle, sb.testGetState())

	b.Shutdown()

	type result struct {
		retryDone RetryDoneFunc
		err       error
	}
	resCh := make(chan result, 1)
	ctx, cancel := context.WithCancel(t.Context())
	t.Cleanup(cancel)

	go func() {
		rd, err := sb.waitForFailoverEnd(ctx, keyspace, shard, nil, failoverErr)
		resCh <- result{rd, err}
	}()

	select {
	case r := <-resCh:
		assert.Nil(t, r.retryDone, "waitForFailoverEnd should return nil RetryDoneFunc after Buffer.Shutdown")
		assert.NoError(t, r.err, "waitForFailoverEnd should return nil error after Buffer.Shutdown")
	case <-time.After(200 * time.Millisecond):
		t.Fatal("waitForFailoverEnd did not return promptly after Buffer.Shutdown — " +
			"it entered the buffering slow path and is blocked in sb.wait(), " +
			"with a timeout-thread goroutine now running unsupervised")
	}

	assert.Equal(t, stateIdle, sb.testGetState(),
		"sb transitioned out of stateIdle after Buffer.Shutdown returned")
}
```

Expected: test passes (post-shutdown call is a no-op).
Actual: test fails — `Starting buffering for shard: ks1/0 …` logs _after_ `Shutdown()` returned, and the goroutine blocks in `sb.wait()` for up to `MaxFailoverDuration`.

### Binary Version

```sh
Reproduces on main at 87a57500d5 and on #19827 HEAD at 12cb13afee.
```

### Operating System and Environment details

```sh
Code-level bug; OS-independent. Reproduced on Linux 6.17 / arm64.
```

### Log Fragments

```sh
Starting buffering for shard: ks1/0 (window: 10s, size: 10, max failover duration: 500ms) (A failover was detected by this seen error: ...)
    shutdown_race_repro_test.go:67: waitForFailoverEnd did not return promptly after Buffer.Shutdown — it entered the buffering slow path and is blocked in sb.wait(), with a timeout-thread goroutine now running unsupervised
--- FAIL: TestShutdownRace_LateWaitForFailoverEnd (0.20s)
```

### Suggested fix direction

In `shardBuffer.waitForFailoverEnd`, after taking `sb.mu.Lock()` on the slow path, check `sb.buf.stopped.Load()` and return `(nil, nil)` if set. This synchronizes with `Buffer.shutdown`'s `Range` (which takes `sb.mu` via `sb.shutdown()`): once `Shutdown` returns, any subsequent `waitForFailoverEnd` observing `stopped=true` bails before calling `startBufferingLocked`.

--- a/go/vt/vtgate/buffer/buffer_test.go
+++ b/go/vt/vtgate/buffer/buffer_test.go
@@ -26,6 +26,7 @@ import (
 	"time"
 
 	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
 
 	"vitess.io/vitess/go/vt/topo/topoproto"
 	"vitess.io/vitess/go/vt/vterrors"
@@ -820,6 +821,50 @@ func testShutdown1(t *testing.T, fail failover) {
 	if err := waitForPoolSlots(b, cfg.Size); err != nil {
 		t.Fatal(err)
 	}
+}
+
+func TestShutdown_WaitForFailoverEndAfterShutdownIsNoop(t *testing.T) {
+	cfg := NewDefaultConfig()
+	cfg.Enabled = true
+	cfg.MaxFailoverDuration = 500 * time.Millisecond
+	b := New(cfg)
+
+	sb := b.getOrCreateBuffer(keyspace, shard)
+	require.NotNil(t, sb)
+	require.Equal(t, stateIdle, sb.testGetState())
+
+	b.Shutdown()
+
+	type result struct {
+		retryDone RetryDoneFunc
+		err       error
+	}
+	resCh := make(chan result, 1)
+	ctx, cancel := context.WithCancel(t.Context())
+	t.Cleanup(cancel)
+
+	go func() {
+		rd, err := sb.waitForFailoverEnd(ctx, keyspace, shard, nil, failoverErr)
+		resCh <- result{rd, err}
+	}()
+
+	var r result
+	require.Eventually(t, func() bool {
+		select {
+		case r = <-resCh:
+			if r.retryDone != nil {
+				r.retryDone()
+			}
+			return true
+		default:
+			return false
+		}
+	}, 30*time.Second, 10*time.Millisecond)
+
+	assert.Equal(t, stateIdle, sb.testGetState(),
+		"sb transitioned out of stateIdle after Buffer.Shutdown returned")
+	assert.Nil(t, r.retryDone, "waitForFailoverEnd should return nil RetryDoneFunc after Buffer.Shutdown")
+	assert.NoError(t, r.err, "waitForFailoverEnd should return nil error after Buffer.Shutdown")
 }
 
 func TestParallelRangeIndex(t *testing.T) {

--- a/go/vt/vtgate/buffer/shard_buffer.go
+++ b/go/vt/vtgate/buffer/shard_buffer.go
@@ -166,6 +166,11 @@ func (sb *shardBuffer) waitForFailoverEnd(ctx context.Context, keyspace, shard s
 
 	// Slow path: acquire lock for state transitions or to enqueue a request.
 	sb.mu.Lock()
+	// Do not start a new buffering cycle after Buffer.Shutdown() has begun.
+	if sb.buf.stopped.Load() {
+		sb.mu.Unlock()
+		return nil, nil
+	}
 	// Re-check state because it could have changed in the meantime.
 	if !sb.shouldBufferLocked(failoverDetected) {
 		// Buffering no longer required. Return early.

